File: gis/src/commonMain/kotlin/jetbrains/gis/geoprotocol/GeometryUtil.java
package jetbrains.gis.geoprotocol;

import com.google.common.collect.Iterators;
import jetbrains.datalore.base.projectionGeometry.LineString;
import jetbrains.datalore.base.projectionGeometry.Multipolygon;
import jetbrains.ocelot.base.geometry.DoubleRectangle;
import jetbrains.ocelot.base.geometry.DoubleRectangles;

import java.util.List;

public final class GeometryUtil {
  public static DoubleRectangle bbox(Multipolygon multipolygon) {
    List<DoubleRectangle> rects = multipolygon.getLimits();
    if (rects.isEmpty()) {
      return null;
    }

    return DoubleRectangles.boundingBox(() -> Iterators.concat(
        Iterators.transform(rects.iterator(), it -> it.origin),
        Iterators.transform(rects.iterator(), it -> it.origin.add(it.dimension))
    ));
  }

  public static LineString asLineString(Geometry geometry) {
    return new LineString(geometry.asMultipolygon().get(0).get(0));
  }

  private GeometryUtil() {
  }
}

File: gis/src/commonMain/kotlin/jetbrains/gis/geoprotocol/GeometryUtil.kt
/*
 * Copyright (c) 2019. JetBrains s.r.o.
 * Use of this source code is governed by the MIT license that can be found in the LICENSE file.
 */

package jetbrains.gis.geoprotocol

import jetbrains.datalore.base.typedGeometry.*

object GeometryUtil {
    fun <TypeT> bbox(multipolygon: MultiPolygon<TypeT>): Rect<TypeT>? {
        val rects = multipolygon.limit()
        return if (rects.isEmpty()) {
            null
        } else {
            sequenceOf(
                rects.asSequence().map { it.origin },
                rects.asSequence().map { it.origin + it.dimension }
            )
                .flatten()
                .asIterable()
                .boundingBox()
        }
    }

    fun <TypeT> asLineString(geometry: MultiPolygon<TypeT>): LineString<TypeT> {
        return LineString(geometry[0][0])
    }
}
